### ISSUE TITLE
redundant_closure must not expand macro in suggestion

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#6850]
Lint name: redundant_closure

````
warning: redundant closure found
   --> src/iter/test.rs:121:13
    |
121 |             || vec![],
    |             ^^^^^^^^^ help: remove closure as shown: `$crate::vec::Vec::new`
    |
    = note: `#[warn(clippy::redundant_closure)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
````

This is obviously not going to work with `$crate::vec...`

The suggestion should just say `Vec::new()`

How to fix: find out where the lint generates the snippet and use `snippet_with_macro_callsite` instead of just `snippet`.

( this type of bug popped up a couple of times already https://github.com/rust-lang/rust-clippy/pull/6801#issuecomment-787063756
maybe we should lint for that?)

**Comment by mgacek8** [comment]
If we go here: https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/eta_reduction.rs#L111 and change that `snipped_opt` to `snippet_with_macro_callsite`, then clippy would suggest the following:
```rust
+error: redundant closure found
+  --> $DIR/eta.rs:207:42
+   |
+LL |     let _ : Option<Vec<i32>> = true.then(||vec![]);
+   |                                          ^^^^^^^^ help: remove closure as shown: `vec![]`
```
which causes the compiler error.

We can strip the `$crate::vec` prefix, but it will work only for the `vec![]` macro. 
This issue is similar: https://github.com/rust-lang/rust-clippy/issues/6732 but with the `quote!()` macro
Maybe there is no sense to trigger redundant closure lint for macros ?

**Comment by camsteffen** [comment]
In the general case, `|| macro!()` should not fire `redundant_closure` since the suggestion will force the user to inline the macro and expose its implementation. But `vec![]` -> `Vec::new` is a special case. So that special case can use a hard-coded suggestion or (better) obtain the minimal path to `Vec` depending on whether it is imported.
